### ISSUE TITLE
Min/max of CartesianIndex are implicitly "vectorized" and conflict with ordering

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/julia#31208]
I hesitate to open this issue because gosh this behavior is so very useful.  But it's probably worth thinking about a better name/idiom here.  In short `min` and `max` for CartesianIndex break the function's contract: they don't return one of the arguments.

```julia
julia> min(CartesianIndex(5, 1), CartesianIndex(1, 5))
CartesianIndex(1, 1)
```

Of course this means that `minimum` also returns an element that is different from `first(sort(idxs))` (where `idxs` is a vector of cartesian indices).

While strange, it is a _hugely_ helpful behavior — e.g., to select the rectangular extents of a found object, you can just do: `idxs = findall(...); region = minimum(idxs):maximum(idxs)`.  I similarly don't really want to change sorting of cartesian indices — it'll put them in column-major order for you!  So if we change this, we'll need to find a new way of spelling `min`/`minimum`/`max`/`maximum`.

One possible alternative is broadcasting: it's currently acting as though we've broadcast `min` across the elements.  We don't support broadcasting over CartesianIndex, but perhaps we should implement tuple-like broadcasting semantics:

```julia
julia> Broadcast.broadcastable(c::CartesianIndex) = c.I

julia> min.(CartesianIndex(5, 1), CartesianIndex(1, 5))
(1, 1) # would need a style and tuple-like implementation to preserve the type
```

That would probably also be helpful in other situations, but it's of no help for `minimum` and `maximum`.

**Comment by jw3126** [comment]
AFAICT the operation `op(CartesianIndex(5, 1), CartesianIndex(1, 5)) = CartesianIndex(1, 1)` is is the infimum with respect to the order `isless` on `CartesianIndex`. So one option would be to introduce functions `inf`, `sup`, `infimum`, `supremum` that behave on `CartesianIndex` like `min, max, minimum, maximum` does currently.

**Comment by martinholters** [comment]
```julia
julia> isless(CartesianIndex(5, 1), CartesianIndex(1, 5))
true
```
so shouldn't the infimum just be `CartesianIndex(5, 1)` for these two?

**Comment by jw3126** [comment]
Sorry you are right. Now I am confused about the order on `CartesianIndex`. I thought it was the product order, but your example shows it is not. The current `minimum` is infimum with respect to the product order.

**Comment by mbauman** [comment]
Right, ordering on CartesianIndex is defined such that if you sort them, they go in column major order.

**Comment by Seelengrab** [comment]
I think `CartesianIndex` should not have an ordering, similar to how `Complex` doesn't (Ref https://github.com/JuliaLang/julia/issues/42060). Recovering the extent of rectangles you mentioned could be done/finished by implementing https://github.com/JuliaLang/julia/issues/43336, which I haven't gotten to yet.

**Comment by goretkin** [comment]
`Complex` doesn't have an ordering because you can multiply and add them, and there is no consistent ordering.
You cannot multiply and add `CartesianIndex` values, so the same issue does not arise.

Note also that `Tuple` and `AbstractVector` are ordered.

**Comment by Seelengrab** [comment]
Not sure what you mean with you cannot multiply and add `CartesianIndex` values, that's what I've been doing for the past few years now:

```
julia> CartesianIndex(1,1) * 5 + CartesianIndex(2,2)
CartesianIndex(7, 7)
```

`CartesianIndex` behaves like a mathematical vector.

**Comment by goretkin** [comment]
You're right that they behave like vectors, with scalar multiplication and addition. I stand corrected. Part of the distinction that I meant to draw, however, is that there is no `*(::CartesianIndex, ::CartesianIndex)`.

**Comment by rapus95** [comment]
just ran into this when I wanted to find the first index (memory-order-sensible) of my index set. This was wildly unexpected for me and feels inconsistent to otherwise clear semantics. (CartesianIndex purposefully doesn't implement vectorization/iteration, but this blindly applies it)

Re my problem: Is this the idiomatic way then?
`partialsortperm(idcs,1)`

**Comment by adienes** [comment]
posterity: #59078 noted that this also breaks the correspondence between `min` / `max` and `clamp`

if [triage's recommendation](https://github.com/JuliaLang/julia/pull/35964#issuecomment-2625357745) to redefine `clamp(x, lo, hi) = min(max(lo, x), hi)` is adopted, the behavior here would change
